man/capfilter.1
.TH capfilter 1 "18 Jun 2011" "BTH" "Measurement Area Manual"
.SH NAME
capfilter \- Filter packages in a capfile.
.SH SYNOPSIS
.nf
.B capfilter [-i \fIFILE\fP] [-o \fIFILE\fP] [\fIOPTIONS...\fP]
.SH DESCRIPTION
.BR capfilter
is used to filter packages in an existing capture and store the result in a new
file. \fIcapfilter\fP without a filter can be used to convert a capture from an
old format to the latest.
.TP
\fB\-p\fR, \fB\-\-packets\fR=\fIN\fR
Stop capture after \fIN\fP packets has been recevied.
.TP
\fB\-m\fR, \fB\-\-matched\fR=\fIN\fR
Stop capture after \fIN\fP packets has been matched.
.TP
\fB\-i\fR, \fB\-\-input\fR=\fIFILE\fR
Read capture from FILE or use \- to read from stdin (default)
.TP
\fB\-o\fR, \fB\-\-output\fR=\fIFILE\fR
Store result into FILE or use \- to write to stdout (default if not a TTY)
.TP
\fB\-r\fR, \fB\-\-rejects\fR=\fIFILE\fR
Store packets rejected by the filter.
.TP
\fB\-v\fR, \fB\-\-invert
Inverts (negates) the filter, i.e. packets that would normally match
will not be discareded and vice-versa.
.TP
\fB\-q\fR, \fB\-\-quiet
Suppress output.
.TP
\fB\-h\fR, \fB\-\-help
Short help text.
.SH FILTER
.TP
\fB\-\-starttime\fR=\fIDATETIME\fR
Discard all packages before DATETIME. See DATE FORMAT for a description of
accepted formats.
.TP
\fB\-\-endtime\fR=\fIDATETIME\fR
Discard all packages including and after DATETIME. See DATE FORMAT for a
description of accepted formats. This will also stop the filtering as no more
packages can match.
.TP
\fB\-\-begin\fR=\fIDATETIME\fR
Alias for --starttime.
.TP
\fB\-\-end\fR=\fIDATETIME\fR
Alias for --endtime.
.TP
\fB\-\-frame\-max\-dt\fR=\fIDATETIME\fR
Discard all packets after DATETIME seconds went since the last packet.
.TP
\fB\-\-iface\fR=\fISTRING\fR
Discard all packages which isn't captured on the interface give by STRING.
.TP
\fB\-\-if\fR=\fISTRING\fR (deprecated)
Alias for \-\-iface.
.TP
\fB\-\-mampid\fR=\fISTRING\fR
Discard all packages not originating from MP with MAMPid STRING.
.TP
\fB\-\-mpid\fR=\fISTRING\fR (deprecated)
Alias for \-\-mampid.
.TP
\fB\-\-eth.vlan\fR=\fITCO[/MASK]\fR
Filter by VLAN Tag Control Information. Discards packets without a VLAN tag.
.TP
\fB\-\-eth.type\fR=\fITYPE[/MASK]\fR
Discards packets not of ethernet protocol \fITYPE\fP which can either be entered as
protocol number or name e.g. "ip" or "arp".
.TP
\fB\-\-eth.src\fR=\fIADDRESS[/MASK]\fR
Discard packets not originating from \fIADDRESS\fP. See ETHERNET ADDRESS for format.
.TP
\fB\-\-eth.dst\fR=\fIADDRESS[/MASK]\fR
Discard packets not destined to \fIADDRESS\fP. See ETHERNET ADDRESS for format.
.TP
\fB\-\-ip.proto\fR=\fIPROTOCOL\fR
Discard packets not using IP protocol \fIPROTOCOL\fP which can either be entered as
protocol number or a valid name from `/etc/protocols`.
.TP
\fB\-\-ip.src\fR=\fIADDRESS\fR[/\fINETMASK\fP]
Discard all packages where source address doesn't match ADDRESS. A mask can be
specified to match a network. Either pass a netmask (e.g. /255.255.255.0) or
CIDR-notation (e.g. /24).
.TP
\fB\-\-ip.dst\fR=\fIADDRESS\fR[/\fINETMASK\fP]
Discard all packages where destination address doesn't match ADDRESS. See
\-\-ip.src for format.
.TP
\fB\-\-tp.sport\fR=\fIPORT[/MASK]\fR
Discard packets not originating from \fIPORT\fP which can either be entered as
protocol number or a valid name from `/etc/services`.
.TP
\fB\-\-tp.dport\fR=\fIPORT[/MASK]\fR
Discard packets not destined to \fIPORT\fP which can either be entered as
protocol number or a valid name from `/etc/services`.
.TP
\fB\-\-tp.port\fR=\fIPORT[/MASK]\fR
Matches packets either to \fBor\fR from \fIPORT\fR,
e.g. \-\-tp.port=80 will match both request and response.
.TP
\fB\-\-caplen\fR=\fIBYTES\fR
Limit the amount of captured bytes to \fIBYTES\fR, truncating packets as needed.
.TP
\fB\-\-filter-mode\fR=\fIMODE\fR
\fIMODE\fP can be set to either OR or AND (default) which changes the operator
when joining the different filters. By setting to OR only one of the filters
has to match while AND all must match.
.TP
\fB\-\-frame-num\fR=\fIRANGE\fR
Discards all packets not in the specified range.
Range can be specified as
.RS
.TP
\[bu] "23" - Match only a specific frame.
.TP
\[bu] "23-59" - Match all frames between 23 and 59 (inclusive)
.TP
\[bu] "-59" - Match all frames up to 59 (inclusive)
.TP
\[bu] "23-" - Match all frames from 23 and onwards (inclusive)
.TP
.RE
Multiple ranges can be joined with comma. E.g. "-5,7,10-" would match packets
1-5, 7 and then from 10 until the end of the stream.
.TP
\fB\-\-bpf\fR=\fIFILTER\fR
Match using a BPF filter. Requires pcap support.
.SH DATE FORMAT
Valid date formats are:
.sp
UNIX timestamp.x
.br
YYYY-MM-DD hh:mm:ss.x
.br
YYYYMMDD hh:mm:ss.x
.br
YYMMDD hh:mm:ss.x
.TP
where x describes picoseconds and can be up to 12 digits.
.SH ETHERNET FORMAT
Ethernet addresses can use an expansion syntax using :: where the space between
the colon is filled with zeroes, e.g. 01::01 is expanded to 01:00:00:00:00:01.
Digits may be separated by colon or dash but does not have to be separated (but
expansion only works with colon). E.g. "ff:ff...", "ff-ff..." and "ffff..." is all
identical.
.SH AUTHOR
Written by David Sveningsson <david.sveningsson@bth.se>.
.SH "SEE ALSO"
mp(1)


NEWS
caputils-0.7.15
---------------

  * pf_ring: compatibility with 6.0

caputils-0.7.14
---------------

  * adding a custom ethernet type for DPMI diagnostic packets.
  * capshow supports many additional protocols.
  * general improvements to output from tools.
  * capdump improved filename generation when using markers.
  * version number always includes git revision if available.
  * new local-only filters for DT and frame numbers.
  * adding generic functions for printing, stateful.
  * pfring always using MTU buffer size instead of hardcoded value.
  * fix issues when doing capture on loopback interfaces.
  * general build fixes.

caputils-0.7.13
---------------

  * pcap2cap improved.
  * Support for BPF filters.
	* capinfo showing more data.

caputils-0.7.12
---------------

  * capfilter silent output.
  * capmerge regression fixes.

caputils-0.7.11
---------------

  * capfilter fixes.

caputils-0.7.10
---------------

  * Additional capshow package decoding.
  * Added stream_peek, stream_num_addresses, timepico_now.
  * stream_copy truncates packages to caplen.

caputils-0.7.9
--------------

  * capinfo output improvements.
  * capdump output improvements.
  * Drops invalid measurement frames.

caputils-0.7.8
--------------

  * Quick bugfixes for PF_RING.

caputils-0.7.7
--------------

  * More statistics and debugging messages.
  * Support for AND/OR filters.
  * Additional functions for developers
  * support for PF_RING when doing ethernet capture.

caputils-0.7.6
--------------

  * caphead_t typedef
  * prototype for stream_from_getopt changed, abi breakage.
  * additional functions for dealing with picotime
  * capshow and capfilter improvements.

caputils-0.7.5
--------------

  * `stream_t` is made a pointer, this will break applications but is easily
    fixed.
  * Additional utility functions.
  * Changed buffering strategy.
  * Added capmarker and capshow.

caputils-0.7.4
--------------

  * Integrating libmarc into caputils.
  * Adding function to add multiple ethernet multicast memberships to a single stream
  * typedef struct stream stream_t

caputils-0.7.3
--------------

	* Moving {,v}logmsg from libmarc here.
	* Major API changes; breaks ABI.
	* Fixed capdump.
	* More man-pages.
	* Fixing compability issues with linux-2.4

caputils-0.7.2
--------------

	* Split filtering code from libcaputils to libfilter.
	* Added two more utilities:
	    - capdump - record a stream
	    - capfilter - filter existing stream
	* Added manpages.

caputils-0.7.0
--------------

Based on the 0.5 branch.
	* Backwards-compability support
	* Numerous API-fixes, the API is *NOT* compatable with earlier versions, but there
	  is only minor fixes required.
	* Numerous bug-fixes, including endianess-fixes.
	* New tools: capinfo, shows information about capfiles.
	* Uses autotools and pkg-config


tools/capfilter.c
/**
 * libcap_utils - DPMI capture utilities
 * Copyright (C) 2003-2013 (see AUTHORS)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <caputils/stream.h>
#include <caputils/filter.h>
#include <caputils/capture.h>
#include <caputils/utils.h>
#include <caputils/version.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <signal.h>
#include <errno.h>
#include <unistd.h>
#include <inttypes.h>

static inline int min(int a, int b){
	return a<b?a:b;
}

static const char* program_name = NULL;
static const char* dst_filename = NULL;
static const char* src_filename = NULL;
static const char* rej_filename = NULL;
static int keep_running = 1;
static int invert = 0;
static int quiet = 0;
static unsigned int max_read = 0;
static unsigned int max_matched = 0;

static const char* shortopts = "p:m:i:o:r:vqh";
static struct option longopts[] = {
	{"packets", required_argument, 0, 'p'},
	{"matched", required_argument, 0, 'm'},
	{"input",   required_argument, 0, 'i'},
	{"output",  required_argument, 0, 'o'},
	{"rejects", required_argument, 0, 'r'},
	{"invert",  no_argument,       0, 'v'},
	{"quiet",   no_argument,       0, 'q'},
	{"help",    no_argument,       0, 'h'},
	{0, 0, 0, 0} /* sentinel */
};

static void show_usage(){
	printf("%s-%s\n", program_name, caputils_version(NULL));
	printf("(C) 2011 david.sveningsson@bth.se\n"
	       "Usage: %s [OPTIONS...] [SRC] [DST] \n"
	       "  -p, --packets=N             Stop after N read packets.\n"
	       "  -m, --matched=N             Stop after N matched packets.\n"
	       "  -i, --input=FILE            read from FILE [default stdin].\n"
	       "  -o, --output=FILE           write to FILE [default stdout].\n"
	       "  -r, --rejects=FILE          write packets not matching to FILE.\n"
	       "  -v, --invert                invert filter.\n"
	       "  -q, --quiet                 suppress output.\n"
	       "  -h, --help                  help (this text).\n"
	       "\n", program_name);
	filter_from_argv_usage();
}

static void handle_sigint(int signum){
	if ( keep_running ){
		fprintf(stderr, "\r%s: got SIGINT, terminating.\n", program_name);
		keep_running = 0;
	} else {
		fprintf(stderr, "\r%s: got SIGINT again, aborting.\n", program_name);
		abort();
	}
}

int main(int argc, char* argv[]){
	/* extract program name from path. e.g. /path/to/MArCd -> MArCd */
	const char* separator = strrchr(argv[0], '/');
	if ( separator ){
		program_name = separator + 1;
	} else {
		program_name = argv[0];
	}

	struct filter filter;
	if ( filter_from_argv(&argc, argv, &filter) != 0 ){
		fprintf(stderr, "%s: Failed to create filter, aborting.\n", program_name);
		exit(1); /* errors already displayed (on stderr) */
	}

	int index = 0;
	int op = 0;
	while ( (op=getopt_long(argc, argv, shortopts, longopts, &index)) != -1 ){
		switch (op){
		case 'p': /* --packets */
			max_read = atoi(optarg);
			break;

		case 'm': /* --matched */
			max_matched = atoi(optarg);
			break;

		case 'o': /* --output */
			dst_filename = optarg;
			break;

		case 'i': /* --input */
			src_filename = optarg;
			break;

		case 'r': /* --rejects */
			rej_filename = optarg;
			break;

		case 'v': /* --invert */
			invert = 1;
			break;

		case 'q': /* --quiet */
			quiet = 1;
			break;

		case 'h': /* --help */
			show_usage();
			exit(0);
		}
	}

	/* use positional arguments unless -i or -o has been specified (for
	 * compatibility with older versions) */
	if ( !(src_filename && dst_filename) ){
		const int nargs = argc-optind;
		switch ( nargs ){
		case 2:
			dst_filename = argv[optind+1];
		case 1:
			src_filename = argv[optind];
		case 0:
			break;
		}
	}

	int ret;
	stream_addr_t addr = STREAM_ADDR_INITIALIZER;
	stream_t src = NULL;
	stream_t dst = NULL;
	stream_t rej = NULL;

	/* ensure not reading/writing capfiles from terminal */
	if ( src_filename == NULL && isatty(STDIN_FILENO) ){
		fprintf(stderr, "%s: Cannot read input from stdin when it is connected to a terminal.\n", program_name);
		fprintf(stderr, "%s: Either specify another destination with --input, use redirection or pipe from another process.\n", program_name);
		exit(1);
	}
	if ( dst_filename == NULL && isatty(STDOUT_FILENO) ){
		fprintf(stderr, "%s: Cannot output to stdout when it is connected to a terminal.\n", program_name);
		fprintf(stderr, "%s: Either specify another destination with --output, use redirection or pipe to another process.\n", program_name);
		exit(1);
	}

	/* defaults */
	src_filename = src_filename ? src_filename : "/dev/stdin";
	dst_filename = dst_filename ? dst_filename : "/dev/stdout";

	/* open source */
	stream_addr_str(&addr, src_filename, 0);
	if ( (ret=stream_open(&src, &addr, NULL, 0)) != 0 ){
		fprintf(stderr, "%s: failed to open input `%s': %s\n", program_name, src_filename, caputils_error_string(ret));
		return 1;
	}

	/* open destination */
	stream_addr_str(&addr, dst_filename, 0);
	if ( (ret=stream_create(&dst, &addr, NULL, "CONV", "capfilter" VERSION " filtered stream")) != 0 ){
		fprintf(stderr, "%s: failed to open output `%s': %s\n", program_name, dst_filename, caputils_error_string(ret));
		return 1;
	}

	/* open rejects */
	if ( rej_filename ){
		stream_addr_str(&addr, rej_filename, 0);
		if ( (ret=stream_create(&rej, &addr, NULL, "CONV", "capfilter" VERSION " filtered stream")) != 0 ){
			fprintf(stderr, "%s: failed to open rejects `%s': %s\n", program_name, rej_filename, caputils_error_string(ret));
			return 1;
		}
	}

	/* handle signals */
	signal(SIGINT, handle_sigint);

	uint64_t matched = 0;
	const struct stream_stat* stats = stream_get_stat(src);
	while ( keep_running ){
		caphead_t cp;
		struct timeval tv = {1,0};
		switch ( (ret=stream_read(src, &cp, NULL, &tv)) ){
		case EAGAIN: /* timeout */
			continue;

		case 0: /* success */
			break;

		default: /* error */
			keep_running = 0;
			continue;
		}

		/* decide what to do with the packet */
		stream_t target = 0;
		const int match = filter_match(&filter, cp->payload, cp);
		const int post_match = invert ? (1-match) : match;
		if ( post_match ){
			target = dst;
			matched++;
		} else if ( rej ){
			target = rej;
		}

		/* truncate if requested */
		if ( filter.caplen != (unsigned int)-1 ){
			cp->caplen = min(filter.caplen, cp->caplen);
		}

		/* copy packet */
		if ( target && (ret=stream_copy(target, cp)) != 0 ){
			fprintf(stderr, "%s: stream_copy() returned %d: %s\n", program_name, ret, caputils_error_string(ret));
			keep_running = 0;
		}

		if ( (max_read > 0 && stats->read >= max_read) || (max_matched > 0 && matched >= max_matched) ){
			/* Read enough pkts lets break. */
			break;
		}
	}

	if ( !quiet ){
		fprintf(stderr, "%s: There was a total of %'"PRIu64" packets read.\n", program_name, stats->read);
		fprintf(stderr, "%s: There was a total of %'"PRIu64" packets matched.\n", program_name, matched);
	}

	filter_close(&filter);
	stream_close(src);
	stream_close(dst);
	stream_close(rej);
	stream_addr_reset(&addr);

	if ( ret != 0 && ret != -1 ){
		fprintf(stderr, "%s: stream_read() returned %d: %s\n", program_name, ret, caputils_error_string(ret));
		return 1;
	}

	return 0;
}
